devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### channel_create
+```c
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create, THANDLE(THREADPOOL), threadpool);
+```
+
+`channel_create` creates the channel and returns it.
+
+**SRS_CHANNEL_43_077: [** If `threadpool` is `NULL`, `channel_create` shall fail and return `NULL`. **]**
+
+**SRS_CHANNEL_43_001: [** `channel_create` shall create a `CHANNEL` object by calling `THANDLE_MALLOC` with `channel_dispose` as `dispose`. **]**
+
+**SRS_CHANNEL_43_078: [** `channel_create` shall create a `CHANNEL_INTERNAL` object by calling `THANDLE_MALLOC`.**]**
+
+**SRS_CHANNEL_43_079: [** `channel_create` shall store the created `THANDLE(CHANNEL_INTERNAL)` in the `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_080: [** `channel_create` shall initialize the `CHANNEL_INTERNAL` with the given `threadpool`. **]**
+
+**SRS_CHANNEL_43_098: [** `channel_create` shall call `srw_lock_create`. **]**
+
+**SRS_CHANNEL_43_081: [** `channel_create` shall call `sm_create`. **]**
+
+**SRS_CHANNEL_43_082: [** `channel_create` shall call `sm_open_begin`. **]**
+
+**SRS_CHANNEL_43_083: [** `channel_create` shall call `sm_open_end`. **]**
+
+**SRS_CHANNEL_43_084: [** `channel_create` shall call `DList_InitializeListHead`. **]**
+
+**SRS_CHANNEL_43_085: [** `channel_create` shall call `interlocked_exchange` to set the `state` to `CHANNEL_EMPTY`. **]**
+
+**SRS_CHANNEL_43_086: [** `channel_create` shall succeed and return the created `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_002: [** If there are any failures, `channel_create` shall fail and return `NULL`. **]**
+
+
+### channel_dispose
+```c
+    static void channel_dispose(CHANNEL* channel);
+```
+
+`channel_dispose` disposes the given `channel`.
+
+**SRS_CHANNEL_43_087: [** `channel_dispose` disposes the given `channel`. **]**
+
+**SRS_CHANNEL_43_088: [** `channel_dispose` shall `sm_close_begin_with_cb` with `channel_close` as the callback. **]**
+
+**SRS_CHANNEL_43_089: [** `channel_dispose` shall `sm_close_end`. **]**
+
+**SRS_CHANNEL_43_090: [** `channel_dispose` shall `sm_destroy`. **]**
+
+**SRS_CHANNEL_43_099: [** `channel_dispose` shall call `srw_lock_destroy`. **]**
+
+**SRS_CHANNEL_43_091: [** `channel_dispose` shall release the reference to `THANDLE(THREADPOOL)`. **]**
+
+**SRS_CHANNEL_43_092: [** `channel_dispose` shall release the reference to `THANDLE(CHANNEL_INTERNAL)`. **]**
+
+
+### channel_close
+```c
+    static void channel_close(void* context);
+```
+
+`channel_close` is the callback passed to `sm_close_begin_with_cb` by `channel_dispose`. It abandons all pending operations
```

Comment:
>It abandons all pending operations [](http://example.com/codeflow?start=89&length=34)

How does it abandon operations for which both the pull and the push have been called, and the OP was given to threadpool_execute_work and the "execute work" didn't happen yet?

Reply:
Seeing as "43_096" says it will iterate over something (it doesn't have it, was given to threadpool_schedule_work).

Reply:
Those operations will not be abandoned. Only the ones still sitting in the queue will be abandoned. Once an OP has been given to threadpool_schedule_work, it cannot be abandoned or cancelled.